### PR TITLE
939/unset index drop

### PR DESCRIPTION
**Pull request by afares97**

### CODE REVIEW

File: static_frame/core/frame.py [hunk 1]
```diff
@@ -6657,13 +6658,18 @@ def unset_index(self, *,
...
-                # No recache is needed as it's not possible for an index to be GO
-                yield from self._index._blocks._blocks # type: ignore
+            if not drop:
+                if self._index.ndim == 1:
+                    yield self._index.values
+                else:
+                    # No recache is needed as it's not possible for an index to be GO
+                    yield from self._index._blocks._blocks # type: ignore
             for b in self._blocks._blocks:
                 yield b
```
Comment (flexatone) [inline static_frame/core/frame.py:6674]: Not part of your changes but this is better as:
```suggestion
            yield from self._blocks._blocks
```

File: static_frame/core/frame.py [hunk 2]
```diff
@@ -6645,6 +6645,7 @@ def set_index_hierarchy(self,
 
     def unset_index(self, *,
             names: tp.Sequence[TLabel] = (),
+            drop: bool = False,
             # index_column_first: tp.Optional[tp.Union[int, str]] = 0,
```
Comment (flexatone) [inline static_frame/core/frame.py:6649]: Not part of your changes but we can remove this now.

File: static_frame/core/frame.py (hunk 1 above)
Comment (afares97) [reply static_frame/core/frame.py:6674]: Yes much cleaner. Will commit suggestion.

File: static_frame/core/frame.py [hunk 3]
```diff
@@ -6657,68 +6657,78 @@ def unset_index(self, *,
             consolidate_blocks:
             columns_constructors:
         '''
+        # disallows specifying names with 'drop=True'
+        if drop and names:
```
Comment (flexatone) [inline static_frame/core/frame.py:6661]: Subtle but by doing this `is` check we implicitly check that the drop argument is a Boolean; that way another truthy value (like `'keep'`!) cannot be used.

```suggestion
        if drop is True and names:
```

File: static_frame/core/frame.py [hunk 4]
```diff
@@ -6657,68 +6657,78 @@ def unset_index(self, *,
...
         else:
             block_gen = blocks
 
-        if not names:
-            names = self._index.names
+        # if not names:
+        #     names = self._index.names
+        if drop:
+            # When dropping the index, keep the existing columns without adding index names
+            columns, own_columns = self._columns, True
```
Comment (flexatone) [inline static_frame/core/frame.py:6684]: We can only own self._columns if this is not a `FrameGO`, i.e., it is purely immutable. We can substitute, the `STATIC` Boolean attr:
```suggestion
            columns, own_columns = self._columns, self.STATIC
```
Please add a test that specifically creates a `FrameGO`, calls this method, then grows (mutating columns) and then ensures the `Frame` resulting from this call does not have its columns mutated.

**Review by flexatone — CHANGES_REQUESTED** [review]
Just a few additional changes.

File: static_frame/core/frame.py (hunk 4 above)
Comment (afares97) [reply static_frame/core/frame.py:6684]: Roger that.

File: static_frame/core/frame.py (hunk 3 above)
Comment (afares97) [reply static_frame/core/frame.py:6661]: Ah nice.

**Review by flexatone — APPROVED** [review]

**Comment by flexatone** [comment]
@afares97 :  I corrected the typing issue; we just needed to a forward declaration show that the value is `None | IndexBase`.
I forgot to mention: we should as add `drop` to `unset_columns`: are up for that?

**Comment by afares97** [comment]
> @afares97 : I corrected the typing issue; we just needed to a forward declaration show that the value is `None | IndexBase`. I forgot to mention: we should as add `drop` to `unset_columns`: are up for that?

Definitely.

**Comment by afares97** [comment]
Thank you for fixing the typing issue. I have added the drop parameter to unset_columns. Will soon add similar tests to ensure correct logic implementation.

File: static_frame/core/frame.py [hunk 5]
```diff
@@ -6877,63 +6886,67 @@ def unset_columns(self, *,
...
+        if drop is True and names:
+            raise RuntimeError("The `names` parameter cannot be used with `drop=True` because the column labels will not be included in the resulting Frame.")
 
-        # columns blocks are oriented as "rows" here, and might have different types per row; when moved on to the frame, types will have to be consolidated "vertically", meaning there is little chance of consolidation. A maximal decomposition might give a chance, but each ultimate column would have to be re-evaluated, and that would be expense.
+        if not names and drop is False:
+            names = self._columns.names
 
-        blocks = TypeBlocks.from_blocks(
+        if drop is True:
+            index, own_index, blocks = self._index, self.STATIC, self._blocks
```
Comment (flexatone) [inline static_frame/core/frame.py:6896]: As Index on the rows are always immutable, here you can assume owning the index.
```suggestion
            index, own_index, blocks = self._index, True, self._blocks
```

**Review by flexatone — CHANGES_REQUESTED** [review]
Looks good; just one thing to change and we should be ready.

**Comment by flexatone** [comment]
@afares97 : note the quality check failure:
```
static_frame/core/frame.py:6944: error: Incompatible types in assignment (expression has type "IndexBase | None", variable has type "IndexBase")  [assignment]
```
Is the same as before: we need to add a pre-usage variable type annotation like this, but for `index` now:
https://github.com/static-frame/static-frame/pull/996/commits/3140904fbb0e84fc77d6bbb1e2f6e36c69f6ae0c

**Comment by afares97** [comment]
Changes are in and tests ran successfully. Thanks for the help.

**Comment by flexatone** [comment]
@afares97 : a few quality things to check on:
https://github.com/static-frame/static-frame/actions/runs/11920048581/job/33221674742?pr=996

**Comment by afares97** [comment]
> @afares97 : a few quality things to check on: https://github.com/static-frame/static-frame/actions/runs/11920048581/job/33221674742?pr=996

Done. I hope.

**Review by flexatone — APPROVED** [review]

**Comment by flexatone** [comment]
All good! Thanks for the contribution, @afares97

**Pull request merged**